### ISSUE TITLE
Confidence loss

### ISSUE THREAD

**Issue by koala7580** [eriklindernoren/PyTorch-YOLOv3#225]
In utils.py line322 
`# Compute label correctness and iou at best anchor
    
    class_mask[b, best_n, gj, gi] = (pred_cls[b, best_n, gj, gi].argmax(-1) == target_labels).float()
    iou_scores[b, best_n, gj, gi] = bbox_iou(pred_boxes[b, best_n, gj, gi], target_boxes, x1y1x2y2=False)
    tconf = obj_mask.float()
`
During training, the confidence of the target should be obj_conf * IoU_best, here the obj_conf is 1, but I didn't see the relationship about confidence and IoU score. Or I am wrong?

**Comment by xiayanqian** [comment]
Well,the value of obj_conf during training is always 1 or 0.99,and the result of detect.py the conf is also 1.maybe we have the same problem.
![image](https://user-images.githubusercontent.com/42026524/62793649-9f657300-bb04-11e9-8681-d4267e201e42.png)

**Comment by xiayanqian** [comment]
Me again.
The confidence of detect.py is cls_conf,not obj_conf,see https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/47b7c912877ca69db35b8af3a38d6522681b3bb3/detect.py#L115

In the models.py,I changed 
https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/47b7c912877ca69db35b8af3a38d6522681b3bb3/models.py#L196
to
`loss_conf_obj = self.bce_loss(pred_conf[obj_mask]*iou_scores[obj_mask], tconf[obj_mask])`
also changed
https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/47b7c912877ca69db35b8af3a38d6522681b3bb3/models.py#L204
to
`conf_obj = (pred_conf[obj_mask]*iou_scores[obj_mask]).mean()`
I don't konw it's correct or not,the conf_obj during training is not 1 any more,the loss getting much bigger though.

**Comment by ly0303521** [comment]
> Well,the value of obj_conf during training is always 1 or 0.99,and the result of detect.py the conf is also 1.maybe we have the same problem.
> ![image](https://user-images.githubusercontent.com/42026524/62793649-9f657300-bb04-11e9-8681-d4267e201e42.png)

问题解决了吗?
我现在测试的时后, conf也很高,不靠谱

**Comment by xiayanqian** [comment]
detect.py的conf不是anchor置信度，是类别分类置信度，因此基本就是1和0.99，是正常现象。同时我改loss的方法是错的，网络无法收敛。

**Comment by ly0303521** [comment]
@

> detect.py的conf不是anchor置信度，是类别分类置信度，因此基本就是1和0.99，是正常现象。同时我改loss的方法是错的，网络无法收敛。

嗯嗯，代码里面确实是类别的置信度，我拿一些图片测试，这个工程得到的类别置信度要么非常高，接近1，要么非常低，接近0。但是用darknet训练出来的模型并不会这样

**Comment by Flova** [comment]
Could you provide an English explanation?
